package_policies/R8.md

```diff
@@ -0,0 +1,7 @@
+**R6.** Each package should provide pre-processor macros for version
+comparison (for languages that support it - like c/c++/fortran).
+Using these macros, dependent software can potentially be compatible
+with multiple versions of the package.
+
+For example - PACKAGE_VERSION_EQ(), PACKAGE_VERSION_LT(),
```

Comment:
I think it should be enough to provide:
```
#define PACKAGE_VERSION_MAJOR 4
#define PACKAGE_VERSION_MINOR 1
```
Do we need the complex macros `PACKAGE_VERSION_***()`

Reply:
I think PACKAGE_VERSION_***() is a better API for users. Otherwise user code would have to construct equivalent complex macros [perhaps multiple times] in the user code. So its best to move this complexity into the package [and not burden the user].

It does requires PACKAGE_VERSION_MAJOR PACKAGE_VERSION_MINOR PACKAGE_VERSION_SUBMINOR and perhaps PACKAGE_VERSION_RELEASE [i.e release or development]

Reply:
The way it is written it is not clear what is actually required to fulfill the policy. The 5 macros are sufficient, but if the package follows good backwards compatibility then a single `PACKAGE_VERSION_GE` is sufficient. If the package changes API on major updates only, then all the user needs is `if (PACKAGE_VERSION_MAJOR > X)` which is just as easy as the `PACKAGE_VERSION_GE`.

In my opinion:

The policy should mandates that packages export variables with compile time information about the version. 
```
#define PACKAGE_VERSION_MAJOR X
#define PACKAGE_VERSION_MINOR Y
#define PACKAGE_VERSION_PATCH Z // if applicable
```
If we mandate additional macros, then `PACKAGE_VERSION_GE` is sufficient.

Reply:
For one the proposal is currently in the 'recommended' category. If part of it can go into mandatory category - that would be great!

And secondly, I don't think xsdk mandates backward compatibility [and I'm not sure if any package provides it].

And I think its trival to have a complete set of macros - so I don't see it as a burden.. [once a package  commits to having this feature]

Wrt wording - improvements are welcome

Reply:
Regardless whether the policy is mandatory or recommended, we should all be able to unambiguously answer whether we comply with the policy or not. Also, if a policy becomes mandatory then we should be able to do that with a simple change from **RX** to **MY**.

Here is my take:

For languages that support pre-processor and compile time mactors/variables/methods (e.g., C/C++/Fortran), packages should provide macros that define the version number as well as a method to check whether the current version is at least X.Y.Z.

For example,
```
#define PACKAGE_VERSION_MAJOR X
#define PACKAGE_VERSION_MINOR Y
#define PACKAGE_VERSION_PATCH Z // if applicable

#define PACKAGE_VERSION_GE(Major, Minor, Patch) \
(((PACKAGE_VERSION_MAJOR==Major) && (PACKAGE_VERSION_MINOR==Minor) && (PACKAGE_VERSION_PATCH>=Patch)) || \
        ((PACKAGE_VERSION_MAJOR==Major) && (PACKAGE_VERSION_MINOR>Minor)) || \
        (PACKAGE_VERSION_MAJOR>Major))
```

In C++-17 and above, static-if is also possible
```
namespace Package_Namespace{
constexpr int VERSION_MAJOR = X;
constexpr int VERSION_MINOR = Y;
constexpr int VERSION_PATCH = Z;

auto VERSION_GE(int major, int minor, int patch){
if (((VERSION_MAJOR==Major) && (VERSION_MINOR==Minor) && (VERSION_PATCH>=Patch)) || 
        ((VERSION_MAJOR==Major) && (VERSION_MINOR>Minor)) || 
        (VERSION_MAJOR>Major)){
    return true;
}
return false;
} 
```

Reply:
I agree with @mkstoyanov here. I think that the major, minor, patch macros are sufficient to fulfill the intentions of the policy, which is to provide dependent software the version at compile time. I agree the other macros might be a good practice, but I also agree that a recommended policy should not need to be reworded to become mandatory. Therefore, maybe a better wording would be:

> **R8.** Each package should provide pre-processor macros that allow dependent software to perform compile time comparisons with the current package version (for languages that support it - like c/c++/fortran). Using these macros, dependent software can potentially be compatible with multiple versions of the package.
>
> For example - PACKAGE_VERSION_MAJOR, PACKAGE_VERSION_MINOR, PACKAGE_VERSION_PATCH, PACKAGE_VERSION_GE(major, minor, patch)

This would also allow for packages which dont necessarily follow the semantic versioning format.

Reply:
It may seem unnecessary, but I've seem many packages with something like 
#define XXX_VERSION  "x.y.z" 
which is not usefully for machine processing. So I suggest an additional clarifying sentence such as 

Note that providing a version number in an include file as a string does not satisfy this policy since it is difficult to process automatically at compile time.

Reply:
@BarrySmith makes good point. We should have the note.

Reply:
I agree as well.

Reply:
Hence me mentioning only PACKAGE_VERSION_EQ(), PACKAGE_VERSION_LT() etc API.
We could always suggest an implementation - but I don't think the policy should mandate an implementation.

If policy mandates PACKAGE_VERSION_MAJOR, PACKAGE_VERSION_MINOR, PACKAGE_VERSION_PATCH etc - its mandating a type of implementation - which I don't think is appropriate.

If a package has only  `#define XXX_VERSION "x.y.z"` - then it won;t be able to provide PACKAGE_VERSION_GE() etc API - so won't be compliant.

Also note: different packages might use different names for PACKAGE_VERSION_PATCH.

For ex: PETSc uses PETSC_VERSION_SUBMINOR, hdf5 uses PETSC_VERSION_SUBMINOR.

Also - some packages might have x.y notation. Others x.y.z notation - so I avoided saying: 
`PACKAGE_VERSION_EQ(major,minor,subminor)`

Reply:
@balay I think we don't really have a disagreement. My wording does not require `PACKAGE_VERSION_MAJOR` or `_PATCH` but `macros that define the version number`. I also require `a method to check whether the current version is at least X.Y.Z.`, but I don't say what it should be named. The rest is an example, but is not mandatory API.

Under my wording, Petsc is already compliant. Packages that don't have such macros would find it easier to understand to add `PACKAGE_VERSION_<whatever applicable>` and only one `greater than` macro. Under the old wording I was confused on how many macros I should provide.

We all agree that the string "x.y.z" is not a number and is not compliant.